traits/src/float.rs

```diff
@@ -402,7 +455,34 @@ pub trait Float
     ///
     /// assert!(abs_difference < 1e-10);
     /// ```
-    fn powi(self, n: i32) -> Self;
+    #[inline]
+    fn powi(self, mut exp: i32) -> Self {
+        if exp == 0 { return Self::one() }
+
+        let mut base;
+        if exp < 0 {
+            exp = -exp;
+            base = self.recip();
+        } else {
+            base = self;
+        }
+
+        while exp & 1 == 0 {
+            base = base * base;
+            exp >>= 1;
+        }
+        if exp == 1 { return base }
+
+        let mut acc = base;
+        while exp > 1 {
+            exp >>= 1;
+            base = base * base;
+            if exp & 1 == 1 {
+                acc = acc * base;
+            }
+        }
+        acc
+    }
```

Comment:
Maybe use the generic `num_traits::pow()` once `exp` is positive?  I see your implementation is identical, but we don't need to duplicate code.